File: java_lexer/Lexer.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.StringTokenizer;
import java.util.regex.*;
import java.util.Dictionary;
import java.util.Arrays;
import java.util.Iterator;
import org.apache.regexp.*;

public class Lexer {

    private boolean verbose = false;

    public enum Token {

	// Terminals

	LEFTPAR     	("LEFTPAR"     , new RE("^" + "\\(")),
	RIGHTPAR    	("RIGHTPAR"    , new RE("^" + "\\)")),
	ASSIGN      	("ASSIGN"      , new RE("^" + "=")),
	COMMA       	("COMMA"       , new RE("^" + ",")),
	SEMICOLON   	("SEMICOLON"   , new RE("^" + ";")),
	PLUS        	("PLUS"        , new RE("^" + "\\+")),
	MINUS       	("MINUS"       , new RE("^" + "-")),
	MULTIPLY    	("MULTIPLY"    , new RE("^" + "\\*")),
	MODULO      	("MODULO"      , new RE("^" + "%")),

	// Keywords
	BEGIN       	("BEGIN"       , new RE("^" + "BEGIN")),
	END         	("END"         , new RE("^" + "END")),
	PRINT       	("PRINT"       , new RE("^" + "PRINT")),

	// Non-implemented
	LEFTBRA     	("LEFTBRA"     , new RE("^" + "\\{")),
	RIGHTBRA    	("RIGHTBRA"    , new RE("^" + "\\}")),
	READ        	("READ"        , new RE("^" + "READ")),
	IF          	("IF"          , new RE("^" + "IF")),
	THEN        	("THEN"        , new RE("^" + "THEN")),
	UNTIL       	("UNTIL"       , new RE("^" + "UNTIL")),
	LESSTHAN    	("LESSTHAN"    , new RE("^" + "<")),
	GREATERTHAN 	("GREATERTHAN" , new RE("^" + ">")),
	REPEAT      	("REPEAT"      , new RE("^" + "REPEAT")),
	EQUALTO     	("EQUALTO"     , new RE("^" + "=")),


	// Char classes
	INTNUM      	("INTNUM"      , new RE("^0|[1-9][0-9]*")),
	ID          	("ID"          , new RE("^[_A-Za-z][_A-Za-z0-9]{0,9}"));

	private final String token;
	private final RE regex;
	Token(String token, RE regex){
	    this.token = token;
	    this.regex = regex;
	}

    }

    public Lexer ( ) {
	this.verbose = false;
    }

    public Lexer ( boolean verbose ) {
	this();
	this.setVerbose(verbose);

    }

    public void setVerbose( boolean verbose) {
	this.verbose = verbose;
    }

    private String tokenize( String input_string ) {
	String input = new String(input_string);
	String output_string = new String("");
	String inputln = Integer.toString(input.length());
	boolean changed;

	do{
	    changed = false;
	    for (Token t : Token.values()){
		String old_input = new String(input);
		String[] split = t.regex.split(input);

		if(this.verbose)
		    System.out.printf("%" + inputln +"s | %s  ==> ", input, t.token, t.regex.toString());

		if ( split.length >= 0 && split[0].length() >= 0 && input.length() > split[0].length()){
		    input = new String(input.substring(split[0].length()+1,input.length()));
		    if (false == input.equals(old_input)){
			if(this.verbose)
			    System.out.printf("%s became %s\n", old_input, input);
			changed = true;
			output_string = output_string.trim() + " " + t.token.trim();
			break;
		    }
		}else{
		    if(this.verbose)
			System.out.println();
		}
	    }
	    if(this.verbose)
		System.out.println();
	}while(changed && input.length() > 0);



	return output_string;

    }

    public static void main(String[] args){
	System.out.println("\nStarting up the Lexer...");
	BufferedReader console = new BufferedReader(new InputStreamReader(System.in));


	boolean verbose_output = false;
	String output_file = null;
	String input_file = null;

	String usage = "Usage: java -cp .:regexp.jar Lexer -v(erbose output)" //
	    + " [-i inputfile (default stdin)] [-o outputfile (default stdout)]";

	for(int i = 0; i < args.length; i++){
	    if (args[i].indexOf("-i") == 0 && args[i].length() > 0){
		String tmp = args[i].substring(2,args[i].length());
		if (tmp.length() == 0){
		    try{
			input_file = args[++i];
			continue;
		    }catch(ArrayIndexOutOfBoundsException e){
			System.out.printf("Bad argument for '-i'. \n%s\n",usage);
			System.exit(1);
		    }
		} else {
		    System.out.printf("Unknown argument: %s. \n%s\n", args[i], usage);
		    System.exit(1);
		}
	    }

	    if (args[i].indexOf("-o") == 0 && args[i].length() > 0){
		String tmp = args[i].substring(2,args[i].length());
		if (tmp.length() == 0){
		    try{
			output_file = args[++i];
			continue;
		    }catch(ArrayIndexOutOfBoundsException e){
			System.out.printf("Bad argument for '-o'. \n%s\n",usage);
			System.exit(1);
		    }
		} else {
		    System.out.printf("Unknown argument: %s. \n%s\n", args[i], usage);
		    System.exit(1);
		}


	    }

	    if (args[i].indexOf("-v") == 0 && args[i].length() == "-v".length()){
		verbose_output = true;
		continue;
	    } else {
		    System.out.printf("Unknown argument: %s. \n%s\n", args[i], usage);
		    System.exit(1);
	    }

	    System.out.printf("Unknown argument: %s. \n%s\n", args[i], usage);
	    System.exit(1);
	}

	if (output_file != null){
	    verbose_output = false;
	}
	System.out.printf("verbose: %b, output: %s, input: %s\n", verbose_output, output_file, input_file);

	Lexer output = new Lexer(verbose_output);


	String tmp = "(3*n+xy)";

	System.out.println(tmp + " ==> " + output.tokenize(tmp));



    }
}

File: python_lexer/lexer.py
import sys,os,re
#spth = os.path.abspath("{0}{1}language_specifications".format(os.getcwd(), os.sep))
#sys.path.insert(0,spth)
#from helper_functions import *
#from token_spec import *

tokens = dict( [x for x in {
    # Terminals
    "LEFTPAR"     : "(?ms)" + re.escape("("),
    "RIGHTPAR"    : "(?ms)" + re.escape(")"),
    "ASSIGN"      : "(?ms)" + re.escape(":="),
    "COMMA"       : "(?ms)" + re.escape(","),
    "SEMICOLON"   : "(?ms)" + re.escape(";"),
    "PLUS"        : "(?ms)" + re.escape("+"),
    "MINUS"       : "(?ms)" + re.escape("-"),
    "MULTIPLY"    : "(?ms)" + re.escape("*"),
    "MODULO"      : "(?ms)" + re.escape("%"),

    # Keywords
    "BEGIN"       : "(?ms)" + re.escape("BEGIN"),
    "END"         : "(?ms)" + re.escape("END"),
    "PRINT"       : "(?ms)" + re.escape("PRINT"),

    # Non-implemented
    "LEFTBRA"     : "(?ms)" + re.escape("{"),
    "RIGHTBRA"    : "(?ms)" + re.escape("}"),
    "READ"        : "(?ms)" + re.escape("READ"),
    "IF"          : "(?ms)" + re.escape("IF"),
    "THEN"        : "(?ms)" + re.escape("THEN"),
    "UNTIL"       : "(?ms)" + re.escape("UNTIL"),
    "LESSTHAN"    : "(?ms)" + re.escape("<"),
    "GREATERTHAN" : "(?ms)" + re.escape(">"),
    "REPEAT"      : "(?ms)" + re.escape("REPEAT"),
    "EQUALTO"     : "(?ms)" + re.escape("="),



    "INTNUM"      : "0|[1-9][0-9]*",
    "ID"          : "[_A-Za-z][_A-Za-z0-9]{,9}",


}.iteritems()])

def arr_to_rexp(arr):
    arr = [ "{0}|".format(x.strip()) for x in arr]
    return "(" + arr_to_str( arr, sep="") + ")"
def arr_to_str(arr, sep=" "):
    return reduce(lambda x,y: x+sep+y, [""] + arr).strip()
def usage():
    return "\nUsage: python " + sys.argv[0] + " [input_file.tny] [output_file.tok]"

if len(sys.argv) < 3:
    sys.stderr.write("Missing argument." + usage() + "\n")
    exit(1)
try:
    in_file = open(sys.argv[1]).read()
except IOError:
    sys.stderr.write("Something is messed up with that file. Try again.\n")
    exit(1)

try:
    out_file = open(sys.argv[2], 'w')
except IOError:
    sys.stderr.write("Something is borked, cannot open {0} with write access.\n".format(sys.argv[2]))
    exit(1)

sstr = arr_to_rexp(tokens.values())

split_input = [y.strip() for x in re.split( sstr, in_file) for y in x.split()]

output = []

comment = False
for x in split_input:
    if comment or re.match(tokens["LEFTBRA"], x):
        comment = True
        if not re.match(tokens["RIGHTBRA"],x):
            continue
        comment = False
        continue

    tok = x

    for y in tokens.items():
        match = re.match(y[1], tok, flags=re.I)
        if match:
            tok = y[0]
            break
        #tok = re.sub(y[1], y[0], tok, flags=re.I)
    output += [(x,tok)]

output = [ (x.strip(),y.strip()) for (x,y) in output ]

out_file

for x in output:
    print "{0: >8} : {1}".format(x[0],x[1])

ostr = ""
for x,y in output:
    #out_file.write(x + " ")
    ostr += (x + " ")
#out_file.write("\n")
ostr += "\n"
for x,y in output:
    out_file.write(y + " ")
    ostr += (y + " ")
out_file.write("\n")
#print ostr
out_file.close()
